pygpcca/utils/_utils.py

```diff
@@ -220,10 +220,54 @@ def _sdd(P: np.ndarray) -> np.ndarray:
     return mu
 
 
+def _eigs_slepc(P: spmatrix, k: int, which: "str" = "LR", tol: float = EPS) -> Tuple[np.ndarray, np.ndarray]:
+    from petsc4py import PETSc
+    from slepc4py import SLEPc
+
+    M = PETSc.Mat().create()
+    if not isspmatrix_csr(P):
+        P = csr_matrix(P)
+    M.createAIJ(size=P.shape, csr=(P.indptr, P.indices, P.data))
+
+    E = SLEPc.EPS()
+    E.create()
+    E.setOperators(M)
+    E.setDimensions(k)
+    E.setTolerances(tol=tol)
+    if which == "LR":
+        E.setWhichEigenpairs(E.Which.LARGEST_REAL)
+    elif which == "LM":
+        E.setWhichEigenpairs(E.Which.LARGEST_MAGNITUDE)
+    else:
+        raise NotImplementedError(f"`which={which}` is not implemented.")
+    E.solve()
+
+    nconv = E.getConverged()
+    if nconv < k:
+        raise ValueError(f"Requested `{k}` eigenvalues/vectors, but only `{nconv}` converged.")
+
+    xr, _ = M.getVecs()
+    xi, _ = M.getVecs()
+
+    eigenvalues, eigenvectors = [], []
+    for i in range(nconv):
+        # Get the i-th eigenvalue as computed by solve().
+        eigenvalues.append(E.getEigenpair(i, xr, xi))
+        if eigenvalues[-1].imag != 0.0:
+            eigenvectors.append([complex(r, i) for r, i in zip(xr.getArray(), xi.getArray())])
+        else:
+            eigenvectors.append(list(xr.getArray()))
+
+    return np.asarray(eigenvalues)[:k], np.asarray(eigenvectors)[:k]
+
+
 @stationary_distribution.register(spmatrix)
 def _sds(P: spmatrix) -> np.ndarray:
     # get the top two eigenvalues and vecs so we can check for irreducibility
-    vals, vecs = eigs(P.transpose(), k=2, which="LR", ncv=None)
+    try:
+        vals, vecs = _eigs_slepc(P.T, k=2, which="LR")
+    except ImportError:
+        vals, vecs = eigs(P.T, k=2, which="LR", ncv=None)
```

Comment:
I suggest a modified way of testing for irreducibility @michalk8. Let's use the test we use in Cellrank, i.e. using networkx to check whether there's more than one connected component. That's the standard way of testing for irreducibility and runs in linear time. The test we have here is something I came up with myself and is a bit more shaky - for example, it depends on these thresholds we have to set below. 

If we use networkx, we don't have to compute the top two eigenvalues but only the top (with largest real part) and we can also get rid of the sorting below (`# sort by real part and take the top one`)

Reply:
Okay never mind, @michalk8 convinced me that this is not necessary - we can leave this check as it is.

Reply:
One thing I would add though is adding some regression tests for this. I have some example matrices we can use here.

Reply:
I'd say not to include any more depenencies if not necessary. On the matrix you've provided, it takes ~3 to get the stat. dist.